src/storage/user-data.ts

```diff
@@ -1,15 +1,60 @@
 import { app } from 'electron';
 import fs from 'fs';
-import nodePath from 'path';
+import nodePath from 'node:path';
 import { SupportedPHPVersion, SupportedPHPVersions } from '@php-wasm/universal';
 import * as Sentry from '@sentry/electron/main';
 import * as atomically from 'atomically';
+import lockfile from 'lockfile';
 import { isErrnoException } from 'src/lib/is-errno-exception';
 import { sanitizeUnstructuredData, sanitizeUserpath } from 'src/lib/sanitize-for-logging';
 import { sortSites } from 'src/lib/sort-sites';
-import { getUserDataFilePath } from 'src/storage/paths';
+import { getResourcesPath, getUserDataFilePath } from 'src/storage/paths';
 import type { PersistedUserData, UserData } from 'src/storage/storage-types';
 
+const LOCKFILE_PATH = nodePath.join( getResourcesPath(), 'appdata-v1.lock' );
+
+function lock( options: lockfile.Options ) {
+	return new Promise< void >( ( resolve, reject ) => {
+		lockfile.lock( LOCKFILE_PATH, options, ( err ) => {
+			if ( err ) {
+				reject( err );
+			} else {
+				resolve();
+			}
+		} );
+	} );
+}
+
+export function unlock() {
+	return new Promise< void >( ( resolve, reject ) => {
+		lockfile.unlock( LOCKFILE_PATH, ( err ) => {
+			if ( err ) {
+				reject( err );
+			} else {
+				resolve();
+			}
+		} );
+	} );
+}
+
+type MaybePromise< T > = T | Promise< T >;
+
+export function withAppdataLock< Args extends unknown[] >(
```

Comment:
Could we do something like this

```typescript
export function withAppdataLock< Args extends unknown[], ReturnType = void >(
	fn: ( userData: UserData, ...args: Args ) => MaybePromise< [ UserData, ReturnType ] >
```

This would allow for the flexibility we need, while keeping everything clean.
What do you think?